treeherder/perf/investigation_aids.py

```diff
@@ -0,0 +1,216 @@
+               decision_task_id=None,
+               task_id=None,
+               task=None,
+               input=None,
+               static_action_variables=None) -> str:
+        context = self._defaults({},
+                                 {
+                                    "taskGroupId": decision_task_id,
+                                    "taskId": task_id or None,
+                                    "input": input,
+                                 },
+                                 static_action_variables)
+
+        if action["kind"] == "task":
+            raise NotImplementedError("Unable to submit actions with 'task' kind.")
+
+        if action["kind"] == "hook":
+            hook_payload = jsone.render(action["hookPayload"], context)
+            hook_id, hook_group_id = action["hookId"], action["hookGroupId"]
+
+            decision_task = self.queue.task(decision_task_id)
+            expansion = self.auth.expandScopes({"scopes": decision_task["scopes"]})
+            expression = f"in-tree:hook-action:{hook_group_id}/{hook_id}"
+
+            if not satisfiesExpression(expansion["scopes"], expression):
+                raise RuntimeError(f"Action is misconfigured: decision task's scopes do not satisfy {expression}")
+
+            result = self.hooks.triggerHook(hook_group_id, hook_id, hook_payload)
+            return result["status"]["taskId"]
+
+    def has_credentials(self):
+        return self.client_id and self.access_token
+
+    @staticmethod
+    def _defaults(target, *sources):
+        """mimics lodash's defaults"""
+        for source in sources:
+            for key, value in source.items():
+                target.setdefault(key, value)
+
+        return target
+
+    @staticmethod
+    def _fetch_job(job_id: str):
+        return (Job.objects
+                .prefetch_related('taskcluster_metadata')
+                .get(id=job_id))
+
+    @classmethod
+    def _filter_relevant_actions(cls, actions_json: dict, original_task) -> List:
+        actions = actions_json['actions']
+
+        def collect_relevant_actions(accumulating_actions, action):
+            acc_action_names = [action['name'] for action in accumulating_actions]
+
+            action_of_known_kind = cls._has_known_kind(action)
+            action_is_relevant = action['name'] not in acc_action_names
+            no_context_or_task_to_check = (not len(action['context'])) and (not original_task)
+            task_is_in_context = (original_task and original_task.get('tags') and
+                                  cls._task_in_context(action['context'], original_task['tags']))
+
+            if (action_of_known_kind and action_is_relevant and
+                    (no_context_or_task_to_check or task_is_in_context)):
+                accumulating_actions.append(action)
+            return accumulating_actions
+
+        actions = reduce(collect_relevant_actions, actions, [])
+        return list(actions)
```

Comment:
Looking at this again, in particular about the handling of names (and thinking about the code that will consume this), it might be easier to collect the actions into a dictionary.[1] This makes `_get_actions` below much simpler.

```python
relevant actions = {}
for action in actions_json['actions']:
    action_name = action['name']
    if action_name in relevant_actions:
         continue
    ....
    if <conditions>:
        relevant_actions[action_name] = action
return relevant_actions
```
---
[1] Per the [action spec](https://docs.taskcluster.net/docs/manual/design/conventions/actions/spec) the order is significant, but:
  1) the code already takes the first action for each name
  2) the secondary reason for order, is to control what gets displayed to a user, which is not relevant here
  3) modern versions of python (3.7+) insertion order is guaranteed to be preserved